Add adaptive width style

LemonIllusion opened the pull request:
There was a [request](https://www.sweclockers.com/forum/post/18001453) to make articles as wide as the forum. I have had this functionality in my fork for a while, but because of a tricky condition that I didn't bother fixing, I never submitted a PR for it... until now! The condition is still not fixed, but maybe @SimonAlling has a good idea of how to either figure the entire condition out or make an acceptable workaround.

If anyone wants to try it out without building themselves, I've provided a testing build below. If you have any feedback you can comment here or @ me in the [official Better SweClockers thread](https://www.sweclockers.com/forum/trad/1541641-better-sweclockers-2018).


## [Install testing build beta.1](https://gist.github.com/LemonIllusion/598413618d1859532b63ece94c7ef4d6/raw/c6a09b31969fd08965ee8b49b22f84d4d559ef0a/better-sweclockers_v3.8.0_adaptive-width-beta.1.user.js)
Testing build built from a5f7f00fa757e5c805d3d72bd6e0a34d95f0f143

LemonIllusion commented:
# Beta.2
- Move instead of remove rightmost ad
- Cosmetic change to code: comments above related blocks in CSS
## [Install testing build beta.2](https://gist.github.com/LemonIllusion/6c73c3179a85a1f63c004276cd907a22/raw/e3bd619df90beec1b31af40bc1f160cb49819fe0/better-sweclockers_v3.8.0_adaptive-width-beta.2.user.js)
Testing build built from 7cef88a4c59b455fd6b6bcc55c0b9e0f049ff887

SimonAlling commented on src/operations.ts:
```diff
@@ -242,6 +243,21 @@ const OPERATIONS: ReadonlyArray<Operation> = [
+            not found? Or should it be an IndependentOperation with its own loop checking
+            for the element's existence?
+         */
+        condition: !isReadingForumThread() && Preferences.get(P.advanced._.adaptive_width),
```
Kan du beskriva vad syftet är med att kolla efter `.pushColumn`? Har inte alla artiklar den? Tror för övrigt att `isReadingEditorialContent()` är vad du söker här, snarare än `!isReadingForumThread()`.

Och behövs det egentligen en operation överhuvudtaget; borde det inte gå att implementera hela featuren enbart med CSS?

SimonAlling commented on src/operations.ts:
```diff
@@ -242,6 +243,21 @@ const OPERATIONS: ReadonlyArray<Operation> = [
         action: REPLACE_FOLLOWED_THREADS_LINK,
     }),
+    new DependentOperation({
+        description: "adaptive width: remove space for .pushColumn>.fixed if empty",
```
När är `.pushColumn > .fixed` tom?

SimonAlling replied:
För övrigt är det tänkt att `description` ska passa in grammatiskt i `Could not ...`; det gör den ju inte nu.

LemonIllusion commented:
`.pushColumn>.fixed` blir tom om man kryssar ur tillräckligt mycket i innehållsfiltret. Just nu får man dessutom köra extra CSS för att bli av med SweC 20 år. Det ser rätt konstigt ut att inte göra något åt den tomma ytan och i halv skärm (som är en av de främsta anledningarna från mitt håll att göra en sådan här stil över huvud taget) vill jag verkligen inte bli begränsad till en minikolumn för nyheterna när det finns plats att utnyttja precis bredvid. Tyvärr kan jag inte komma på ett bra sätt att göra det här med enbart CSS.

![](https://i.imgur.com/YHVV3B6.png)
![](https://i.imgur.com/Yo9GSn5.png)

`.pushColumn` finns på bland annat:
- Profilsida
- Marknadsannons
- Sök i arkivet
- Framsida
- Artiklar

Den saknas på bland annat:
- Alla olika sorters undersidor till forumet
- Gallerier
- Marknad
- Guider

Villkoret blir väldigt komplicerat oavsett om man gör white- eller blacklist, så jag hade helst kört en "går det så går det"-lösning som misslyckas i tystnad på sidor som saknar `.pushColumn`.

LemonIllusion commented:
Maybe I should just have the operation in my own fork where I don't really care about having errors in the console. I'm probably the only one who makes use of it anyway. Anyway, I made some changes, let's see where they take us.

# Beta.3
- Use limited whitelist for `.pushColumn>.fixed` removal
It won't work everywhere now, but at least the two most important cases, front page and market listings are working. I could live with this as most other pages where it happens are uncommon to visit ([example](https://www.sweclockers.com/kontakt))
- Fix image in "Guider" not being full width
- Fix description of operation to fit grammatically

## [Install testing build beta.3](https://gist.github.com/LemonIllusion/6c044027344a61fea44bf3ba81944432/raw/eb3285724c5cd27c19908a8dc01aee6b5fc703c4/better-sweclockers_v3.8.0_adaptive-width-beta.3.user.js)
Testing build built from 61e9cac0a8be854a146777d235ee87c5d18b2216

SimonAlling commented:
The one thing I'm completely sure about is that you should not keep this feature to your fork, because it's a very nice feature. :) I think it might be possible to do it with CSS only; otherwise, I think the best solution is to decide whether to apply the style based on the path – i.e. something like `adaptiveWidthStyleMakesSense` in `environment.ts`.

SimonAlling commented:
> `.pushColumn>.fixed` blir tom om man kryssar ur tillräckligt mycket i innehållsfiltret. Just nu får man dessutom köra extra CSS för att bli av med SweC 20 år. Det ser rätt konstigt ut att inte göra något åt den tomma ytan och i halv skärm (som är en av de främsta anledningarna från mitt håll att göra en sådan här stil över huvud taget) vill jag verkligen inte bli begränsad till en minikolumn för nyheterna när det finns plats att utnyttja precis bredvid. Tyvärr kan jag inte komma på ett bra sätt att göra det här med enbart CSS.

Tack, bra förklarat!

> `.pushColumn` finns på bland annat:
> 
> * Profilsida
> * Marknadsannons
> * Sök i arkivet
> * Framsida
> * Artiklar
> 
> Den saknas på bland annat:
> 
> * Alla olika sorters undersidor till forumet
> * Gallerier
> * Marknad
> * Guider
> 
> Villkoret blir väldigt komplicerat oavsett om man gör white- eller blacklist, så jag hade helst kört en "går det så går det"-lösning som misslyckas i tystnad på sidor som saknar `.pushColumn`.
> 
> En annan lösning skulle kunna vara att bara låta `.pushColumn>.fixed` kunna försvinna på framsidan. Jag ska testa och se om det är något jag kan vara nöjd med.

Vi vill dock inte gärna göra den här typen av modifieringar med DOM-operationer, eftersom det innebär att sidan laddas in och _sedan_ ändras bredden på huvudinnehållet. Det ser inte bra ut och är mycket irriterande i synnerhet på långsammare uppkopplingar. Vi bör göra allt för att lösa detta med CSS.

Ska för övrigt nämna att vi ej vill röra `.pushColumn` på sidor som [`/arkiv`](https://sweclockers.com/arkiv), då den är essentiell för funktionaliteten där.

LemonIllusion commented:
Fan vad jag älskar att du är envis! Jag skrev om kolumndelen med flex och strösslade med lite magi och det ser faktiskt ut som att det fungerar med bara CSS nu. Jag vågar inte påstå att jag har någon susning om varför eller hur det funkar, men det kan man klura på någon annan dag.

# Beta.4
- Flex, magic and tiny bit of luck
- Piped the DOM operation to `/dev/null`
- Things might break in new hilarious ways ←feature
- ~~By accident~~definitely on purpose, `.pushColumn>.fixed` is no longer fixed, but I like it better this way (until someone points out where it breaks badly)
Keep an eye on if it grows beyond what it was before (154px inner width, 162px outer width)

## [Install testing build beta.4](https://gist.github.com/LemonIllusion/e7f6964bd024e5cb243784c52d1b0e63/raw/aceb690054d061cf1fb3a753751672a24f43820e/better-sweclockers_v3.8.0_adaptive-width-beta.4.user.js)
Testing build built from b6674e2d2e7318fd89ac45f953f7b903c62298b9

LemonIllusion commented:
First bug:  [`/arkiv`](https://www.sweclockers.com/arkiv), "Mest kommenterade" doesn't fit.
![](https://i.imgur.com/7arUDFr.png)

LemonIllusion commented:
No new bugs have been reported and the new CSS looks like it's working well. Let's get this wheel spinning for real now.
# RC.1
- Fixed width on `.pushColumn>.fixed`, it was fun while it lasted
- Rectified image width on [`/arkiv`](https://www.sweclockers.com/arkiv)

## [Install release candidate 1](https://gist.github.com/LemonIllusion/ab31ace5891224751d4df55a4bcfe78b/raw/e83cfd3935826c69ea785ea96c5a76083f53797c/better-sweclockers_v3.8.0_adaptive-width-rc.1.user.js)
built from b277532001c6ef94bebf20b4c0981167fde52d13

LemonIllusion commented:
Största frågetecknet från min sida är var rutan i inställningarna ska vara. Beta-texten tar jag bort när jag flyttar den till rätt ställe.

SimonAlling commented:
Tycker den ska ligga under _Allmänt_ (`general`), precis efter _Kompakt layout_. Övriga kommentarer och åtgärder får nog vänta till efter Qontinent.

SimonAlling commented:
There are two more things I would like to bring up before merging:

* I think the option name should be more concise. What about _LemonIllusions adaptiva layout_? Or maybe _Adaptiv artikelbredd_?
* The adaptive layout doesn't work very well together with _Bättre rättelsegränssnitt_. (You can see this by hitting _Skicka en rättelse_ to bring up the corrections interface.) However, I don't think this is a major issue, because most people don't submit a lot of corrections, and either of the options can be disabled anyway. Ergo, you don't have to fix this for the PR to be merged.

LemonIllusion commented:
The positioning CSS for the _Bättre rättelsegränssnitt_ looked a bit funky to me and the transform makes the entire box misaligned with the pixel grid resulting in the entire box getting blurry. I will suggest changes in a separate PR and make it work better with the adaptive layout as well.
![](https://i.imgur.com/SHyTciC.png)
![](https://i.imgur.com/ys64XfY.png)
